packages/fuels-programs/src/receipt_parser.rs

```diff
@@ -0,0 +1,318 @@
+use fuel_tx::{ContractId, Receipt};
+use fuels_core::abi_decoder::ABIDecoder;
+use fuels_types::{
+    bech32::Bech32ContractId,
+    error,
+    errors::{Error, Result},
+    param_types::{ParamType, ReturnLocation},
+    Token,
+};
+use itertools::Itertools;
+
+pub struct ReceiptParser {
+    receipts: Vec<Receipt>,
+}
+
+impl ReceiptParser {
+    pub fn new(receipts: &[Receipt]) -> Self {
+        let relevant_receipts: Vec<Receipt> = receipts
+            .iter()
+            .filter(|receipt| {
+                matches!(receipt, Receipt::ReturnData { .. } | Receipt::Return { .. })
+            })
+            .cloned()
+            .collect();
+
+        Self {
+            receipts: relevant_receipts,
+        }
+    }
+
+    /// Based on receipts returned by a script transaction, the contract ID (in the case of a contract call),
+    /// and the output param, parse the values and return them as Token.
+    pub fn parse(
+        &mut self,
+        contract_id: Option<&Bech32ContractId>,
+        output_param: &ParamType,
+    ) -> Result<Token> {
+        let contract_id = contract_id
+            .map(Into::into)
+            // During a script execution, the script's contract id is the **null** contract id
+            .unwrap_or_else(ContractId::zeroed);
+
+        let data = self
+            .extract_raw_data(output_param, &contract_id)
+            .ok_or_else(|| Self::missing_receipts_error(output_param))?;
+
+        ABIDecoder::decode_single(output_param, &data)
+    }
+
+    fn missing_receipts_error(output_param: &ParamType) -> Error {
+        error!(
+            InvalidData,
+            "ReceiptDecoder: failed to find matching receipts entry for {output_param:?}"
+        )
+    }
+
+    fn extract_raw_data(
+        &mut self,
+        output_param: &ParamType,
+        contract_id: &ContractId,
+    ) -> Option<Vec<u8>> {
+        match output_param.get_return_location() {
+            ReturnLocation::ReturnData if output_param.is_vm_heap_type() => {
+                self.extract_return_data_heap(contract_id)
+            }
+            ReturnLocation::ReturnData => self.extract_return_data(contract_id),
+            ReturnLocation::Return => self.extract_return(contract_id),
+        }
+    }
+
+    fn extract_return_data(&mut self, contract_id: &ContractId) -> Option<Vec<u8>> {
+        for (index, receipt) in self.receipts.iter_mut().enumerate() {
+            if let Receipt::ReturnData { id, data, .. } = receipt {
+                if id == contract_id {
+                    let data = std::mem::take(data);
+                    self.receipts.remove(index);
+                    return Some(data);
+                }
+            }
+        }
+        None
+    }
```

Comment:
How about rewriting this into: 

```suggestion
        fn extract_return_data(&mut self, contract_id: &ContractId) -> Option<Vec<u8>> {
        if let Some(index) = self.receipts.iter().position(|receipt| {
            if let Receipt::ReturnData { id, .. } = receipt {
                id == contract_id
            } else {
                false
            }
        }) {
            if let Receipt::ReturnData { data, .. } = self.receipts.remove(index) {
                Some(data)
            } else {
                None
            }
        } else {
            None
        }
    }
```

Correct if I'm wrong, but, I don't _think_ we need `std::mem::take()` here. These changes make the code more concise and expressive, while also improving its performance by minimizing the number of iterations required to find the relevant receipt. Thoughts?

Reply:
@digorithm I feel that with your suggestion, with the first `if` my brain would need a comment to understand right away. But there is bound to be a lot of receipts parsing so I guess the performance can be important. Maybe just add a comment saying `look for ReturnData receipts with the right contract id`?

Reply:
Yeah. I'm not usually super picky with performance regarding the SDK, but receipt parsing can be one area of performance that could matter.

Reply:
The `receipts` are already filtered at this point to only contain `Return` and `ReturnData`. If we ever measure the performance to be sub par we can come back to this, but right now I think the difference is negligible. 

Since @iqdecay mentioned that this alternative would also warrant a comment, I would rather stick with the current version. 

For my brain it also took a moment to digest that the first if is a filtering step.